Fix various bugs and add "floating IP handler" to the OpenStack provider

hcs42 opened the pull request:
This commit contains:
- Fixes of typos in docstrings and comments
- 7 various small bugfixes
- a small feature (add "floating IP handler" to the OpenStack provider)
- a small addition to the DUMMY provider

If I should break up this pull request into several pull requests and/or I should write Jira tickets, just let me know.

hcs42 committed:
Fix import key in EC2 driver

The original code was faulty, because the
ex_find_or_import_keypair_by_key_material function returned a key pair object,
and that object was placed in params['KeyName'], which was later expected to
contain only the key name, not the key pair object.

This fix places only the key name in params['KeyName'].

hcs42 committed:
Add import_key_pair_from_string to DummyNodeDriver

hcs42 committed:
Fix typos in docstrings

hcs42 committed:
Bugfix: fix variable name in EC2 driver

"id" is a built-in function in Python, which caused this error to slip through
the unit tests.

hcs42 committed:
ec2 driver: Make cidr_ips argument mandatory

If the cidr_ips parameter in this function is None, and therefore the
IpPermissions.1.IpRanges.1.CidrIp is not present in the request to the EC2 API,
the request is not performed. (The response does not indicate an error, but the
security rule is not created).

This behaviour is not documented in
http://docs.aws.amazon.com/AWSEC2/latest/APIReference/ApiReference-query-AuthorizeSecurityGroupIngress.html,
but it is reproducable.

The following request didn't work:

    Action=AuthorizeSecurityGroupIngress
    GroupId=sg-a4b44ece
    IpPermissions.1.FromPort=22
    IpPermissions.1.IpProtocol=tcp
    IpPermissions.1.ToPort=22
    AWSAccessKeyId=...
    Signature=...
    SignatureMethod=HmacSHA256
    SignatureVersion=2
    Timestamp=2014-04-16T07%3A43%3A21Z
    Version=2013-10-15

While this one did (the only difference is adding CidrIp):

    Action=AuthorizeSecurityGroupIngress
    GroupId=sg-a4b44ece
    IpPermissions.1.FromPort=22
    IpPermissions.1.IpProtocol=tcp
    IpPermissions.1.ToPort=22
    IpPermissions.1.IpRanges.1.CidrIp=0.0.0.0/0
    AWSAccessKeyId=...
    Signature=...
    SignatureMethod=HmacSHA256
    SignatureVersion=2
    Timestamp=2014-04-16T07%3A43%3A21Z
    Version=2013-10-15

hcs42 committed:
Add "floating IP" functions to the OpenStack provider

The Compute API of the HP Cloud used by Libcloud [1] does not support floating IP
pools, but it does support floating IPs. (To be more precise, one needs to use
floating IPs if he wants to access the instance from outside the cloud.)

So I added floating ip handling methods to the OpenStack_1_1_NodeDriver class.

[1] HP Helion Public Cloud v13.5 Compute Service:
https://docs.hpcloud.com/api/v13/compute/#3.10Extensions

hcs42 committed:
OpenStack/delete security group: Accept ACCEPTED

According to https://wiki.openstack.org/wiki/Os-security-groups#Delete_Security_Group
and my own experiments with the HP cloud, the provider's answer upon success is
not NO_CONTENT (HTTP status code 204) but ACCEPTED (HTTP status code 202).

So I modified to code to accept ACCEPTED. I still kept NO_CONTENT (I would
think it was put there because some provider did return NO_CONTENT).

hcs42 committed:
Fix typo in comment

hcs42 committed:
Fix ec2.import_key_pair_from_string for Python 3

In Python 3, the original code didn't work, because base64key was a byte
string, and it became a value in the `params` dictionary, but that needs
strings as values.

hcs42 committed:
Fix publickey._to_md5_fingerprint for Python 3

In Python 3, the original code didn't work, because there is no `encode`
function for binaries:

    (Python3.3)>>> [x.encode("hex") for x in b"abc"]
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
      File "<stdin>", line 1, in <listcomp>
    AttributeError: 'int' object has no attribute 'encode'

String do have an `encode` method, but it is a different function from the
`encode` in Python 2:

    (Python3.3)>>> [x.encode("hex") for x in "abc"]
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
      File "<stdin>", line 1, in <listcomp>
    LookupError: unknown encoding: hex

So I added a "hexadigits" function to our compatibility module (py3) and used
that in _to_md5_fingerprint.

hcs42 committed:
Fix publickey.get_pubkey_ssh2_fingerprint for Python 3

In Python 3, the original code didn't work, because there we cannot concatenate
a string and a byte string.

The new code produces two binaries in Python 3 and two string in Python 2, so
they can be concatenated in both versions.

Kami commented on libcloud/utils/publickey.py:
```diff
@@ -16,6 +16,10 @@
 import base64
 import hashlib
 
+from Crypto.Util.py3compat import bchr
```
paramiko (and pycrypto) is an optional dependency, but `get_pubkey_openssh_fingerprint` should also work without those libraries so we can't rely on `bchr` function from `pycrypto`.

hcs42 replied:
Is it OK if I add a `bchr` function to our `py3` module?

Kami replied:
Yep

Kami commented on libcloud/utils/py3.py:
```diff
@@ -93,6 +101,11 @@ def dictvalues(d):
     def tostring(node):
         return ET.tostring(node, encoding='unicode')
+
+    def hexadigits(s):
```
It would be nice to add a small test case for this.

hcs42 replied:
Will do.

Kami commented on libcloud/compute/drivers/openstack.py:
```diff
@@ -2168,6 +2168,14 @@ def ex_list_floating_ip_pools(self):
         return self._to_floating_ip_pools(
             self.connection.request('/os-floating-ip-pools').object)
 
+    def ex_create_floating_ip_handler(self):
```
For consistency, please avoid handler approach and add method directly to the driver. I'm not totally sure about the name though since `ex_create_floating_ip` might be confusing since we already have `OpenStack_1_1_FloatingIpPool.create_floating_ip`.

hcs42 replied:
I did it this way so that I can easier reuse the code from `OpenStack_1_1_FloatingIpHandler`.

But that's not a lot of code, and I agree that it would be more convenient to use this functionality without having to create a handler. I will rewrite it and we will see how that looks like.

hcs42 commented:
I updated three commits:
- "Fix publickey._to_md5_fingerprint for Python 3": added a unit test.
- "Fix publickey.get_pubkey_ssh2_fingerprint for Python 3": use our own py3.bchr function.
- "Add "floating IP" functions to the OpenStack provider": instead of using a handler, I added IP address handler methods directory to the driver class. I used the method name you mentioned, I don't find it's confusing.
  
  I had to add a "driver" parameter to the OpenStack_1_1_FloatingIpAddress class so that the object can use it to delete itself. I didn't want to use the "pool" object for this, I felt that that would be more confusing. Suggestions about better solutions are welcome.

Kami commented:
I've squashed the commits, fixed some lint issues and merged your changes into trunk.

Thanks!

Kami commented:
@hcs42 Btw, I know this has already been merged, but what do you think about this approach - https://issues.apache.org/jira/browse/LIBCLOUD-567?focusedCommentId=14012135&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14012135 ?

hcs42 commented:
Thanks! I posted my opinion in the JIRA ticket.
